src/geode/georepresentation/msh_input.cpp

```diff
@@ -348,49 +350,101 @@ namespace
+            build_corners();
+            build_lines();
+            build_surfaces();
+        }
+
+        using boundary_incidences_relations = std::unordered_map< geode::uuid,
+            std::unordered_set< geode::uuid > >;
+
+        void build_topology()
+        {
+            const auto& model_vertices = brep_.unique_vertices();
+
+            boundary_incidences_relations corner_line_relations;
+            boundary_incidences_relations line_surface_relations;
+            boundary_incidences_relations surface_block_relations;
+
+            for( auto uv : geode::Range{ model_vertices.nb_unique_vertices() } )
+            {
+                const auto corners_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Corner3D::component_type_static() );
+                const auto lines_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Line3D::component_type_static() );
+                const auto surfaces_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Surface3D::component_type_static() );
+                const auto blocks_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Block3D::component_type_static() );
+
+                add_potential_relationships(
+                    corners_vertices, lines_vertices, corner_line_relations );
+                add_potential_relationships(
+                    lines_vertices, surfaces_vertices, line_surface_relations );
+                add_potential_relationships( surfaces_vertices, blocks_vertices,
+                    surface_block_relations );
+            }
+            for( auto uv : geode::Range{ model_vertices.nb_unique_vertices() } )
+            {
+                const auto corners_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Corner3D::component_type_static() );
+                const auto lines_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Line3D::component_type_static() );
+                const auto surfaces_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Surface3D::component_type_static() );
+                const auto blocks_vertices =
+                    model_vertices.mesh_component_vertices(
+                        uv, geode::Block3D::component_type_static() );
+
+                filter_potential_relationships(
+                    corners_vertices, lines_vertices, corner_line_relations );
+                filter_potential_relationships(
+                    lines_vertices, surfaces_vertices, line_surface_relations );
+                filter_potential_relationships( surfaces_vertices,
+                    blocks_vertices, surface_block_relations );
+            }
+
```

Comment:
Maybe the builder could be an attribute member. You are using it in different places

Reply:
Done